File: patterns/observers/Observer.kt
/**
 * RESOURCES USED:
 * 
 * This one I watched the entire video, very helpful resource
 * https://www.youtube.com/watch?v=_BpmfnqjgzQ&t=2245s
 * 
 * This one I watched part of it, mainly for understanding how to write with kotlin, as this is my first time.
 * https://www.youtube.com/watch?v=5WIS-GydJmw&t=585s
 * 
 * Additionally, I used stack overflow, kotlin docs and other misc resources that I can't remember
 */
// Interface for the subject(s) -> the one(s) who will be watched
interface Subjects {
  fun subscribe(observer: Observers)
  fun unsubscribe(observer: Observers)
  fun notifyObservers()
  fun setState(state: String?)
  fun getState(): String?
}

// Interface for the observer(s) -> the one(s) who will be watching
interface Observers {
  fun update(subject: Subjects)
}

class Subject : Subjects {
  // creates a list of observers
  private val observersList = mutableListOf<Observers>()
  // creates a default state
  private var state: String? = ""

  // adds an observer to a subject
  override fun subscribe(observer: Observers) {
      if(!observersList.contains(observer)) {
          println("$observer is now subscribed to $this");
          observersList.add(observer)
      } else {
          println("$observer is already subscribed to $this");
      }
  }
  // adds an array of observers to a subject
  fun subscribe(observers: MutableList<Observers>) {
      observers.forEach { observer ->
          subscribe(observer)
      }
  }

  // removes an observer from a subject
  override fun unsubscribe(observer: Observers) {
      // checks if observer is currently subscribed
      if (observersList.contains(observer)) {
          println("$observer is now unsubscribed to $this");
          observersList.remove(observer)
      } else {
          println("$observer is not subscribed to $this");
      }
  }
  // removes an array of observers from a subject
  fun unsubscribe(observers: MutableList<Observers>) {
      observers.forEach { observer ->
          unsubscribe(observer)
      }
  }
  // notifies all observers that the subject is changed
  override fun notifyObservers() {
      observersList.forEach { observer ->
          observer.update(this)
      }
  }

  // sets the state of the subject
  override fun setState(state: String?) {
      this.state = state
  }

  // returns the state of the subject to the observer
  override fun getState(): String? {
      return this.state
  }
}

class Observer : Observers {
  override fun update(subject: Subjects) {
      println("$this Subject Change now: ${subject.getState()}")
  }
}

File: patterns/Behavioral/Observal/Kotlin/Observer.php
<?php
/**
 * RESOURCES USED:
 * I wrote the Kotlin observer pattern and used the same resources
 * 
 * This one I watched the entire video, very helpful resource
 * https://www.youtube.com/watch?v=_BpmfnqjgzQ&t=2245s
 * 
 * Additionally, I used stack overflow and other misc resources that I can't remember
 */
interface Observers
{
  public function __construct();
  public function update($subject);
}

interface Subjects
{
  public function __construct();
  public function subscribe(Observer $observer);
  public function unsubscribe(Observer $observer);
  public function notify();
  public function setData($data);
  public function getData();
}

/** 
 * Description:
 * Use to observe subjects and update only after
 * the subject has changed.
 * 
 * @since 1.0.0
 */
class Observer implements Observers
{
  /**
   * Creates a unique ID for Observers
   * @var $id
   */
  private $id;

  public function __construct()
  {
    $this->id = uniqid('observer_');
  }
  public function update($subject)
  {
   echo "$this->id notified that the subject has changed to ".$subject->getData()."\r\n";
  }
  /**
   * Get ID
   * @return Observer $id
   */
  public function getId()
  {
    return $this->id;
  }
}
/**
 * Description:
 * Used to create subjects that notify observers
 * after they've changed.
 * 
 * @since 1.0.0
 */
class Subject implements Subjects
{
  private $observerList = array();
  private $data;
  private $id;

  /**
   * Creates a unique ID for Subjects
   * @var $id
   */
  public function __construct()
  {
    $this->id = uniqid('subject_');
  }
  /**
   * Subscribes Observer object to Subject object
   * @param Observer
   */
  public function subscribe(Observer $observer)
  {
    if (!in_array($observer, $this->observerList)) {
      $this->observerList[] = $observer;
      echo  $observer->getId()." is now subscribed to $this->id\r\n";
    } else {
      echo $observer->getId()." is already subscribed to $this->id\r\n";
    }
  }

  /**
   * Unsubscribe Observer object from Subject object
   * @param Observer
   */
  public function unsubscribe(Observer $observer)
  {
    $key = array_search($observer, $this->observerList);
    if ($key !== false) {
      unset($this->observerList[$key]);
      echo $observer->getId()." is now unsubscribed from $this->id\r\n";
    } else {
      echo $observer->getId()." is not subscribed to $this->id\r\n";
    }
  }

  /** 
   * Set Subject data
   * @param string
   */
  public function setData($data)
  {
    $this->data = $data;
  }

  /**
   * Send an ping to each observer subscribbed
   */
  public function notify()
  {
      foreach ($this->observerList as $observer) {
          $observer->update($this);
      }
  }

  /**
   * give data to observer
   * @return Subject $data
   */
  public function getData()
  {
    return $this->data;
  }
}
